controllers/volume/v1alpha1/volume_controller.go

```diff
@@ -624,18 +624,68 @@ func (r *Reconciler) createCNSAttachment(
 	}
 
 	if err := r.Create(ctx, attachment); err != nil {
-		// An IsAlreadyExists error is not really expected because the only way we'll do a Create()
-		// is if the attachment wasn't returned in the List() done at the start of the reconcile.
-		// The Client cache might have been stale, so ignore the error, but like the comment in the
-		// caller, we might want to perform an Update() in this case too.
-		if !apiErrors.IsAlreadyExists(err) {
-			return errors.Wrap(err, "Error creating CnsNodeVmAttachment")
+		if apiErrors.IsAlreadyExists(err) {
+			return r.createCNSAttachmentButAlreadyExists(ctx, attachmentName)
 		}
+		return errors.Wrap(err, "Cannot create CnsNodeVmAttachment")
 	}
 
 	return nil
 }
 
+// createCNSAttachmentButAlreadyExists tries to handle various conditions when a CnsNodeVmAttachment
+// unexpected already exists. Usually the existing attachment is for a prior VC VM has been deleted
+// from underneath us, and the replacement will have a different BiosUUID. The CnsNodeVmAttachment
+// workflow was from CNS and has been very cumbersome to us to use in practice, with this controller
+// being more complicated than it really should be for us to try to work around all the edge cases.
+// Note that CNS doesn't reevaluate an attachment at all once Status.Attached=true. We could Update
+// it instead - clearing Attached atomically since it doesn't have a Status subresource - but it
+// should not be treating the Status as the SoT.
+func (r *Reconciler) createCNSAttachmentButAlreadyExists(
+	ctx *context.VolumeContext,
+	attachmentName string) error {
+
+	attachment := &cnsv1alpha1.CnsNodeVmAttachment{}
+	err := r.Client.Get(ctx, client.ObjectKey{Name: attachmentName, Namespace: ctx.VM.Namespace}, attachment)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			// Create failed but now the attachment does not exist. Most likely the attachment got GC'd.
+			// Return an error to force another reconcile to create it.
+			return errors.Errorf("stale client cache for CnsNodeVmAttachment %s that now does not exist. Force re-reconcile", attachmentName)
+		}
+
+		return err
+	}
+
+	if !metav1.IsControlledBy(attachment, ctx.VM) {
+		// This attachment has our expected name but is not owned by us. Most likely, the owning VM
+		// is in the process of being deleted, and the attachment will be GC after the owner is gone.
+		// We just have to wait it out here: we cannot delete an attachment that isn't ours.
+		return errors.Errorf("CnsNodeVmAttachment %s has a different controlling owner", attachmentName)
```

Comment:
Should we consider having a requeue delay to avoid the controller being swarmed by repeated reconciles in these error scenarios (L664 and L657)?

Reply:
In the most common case here, our k8s VM is still the owner of the attachment so our Watch will result in prompt re-reconcile. Otherwise, we already get an exponential backoff for free from ctrl-runtime when returning an error; I'm not going to do anything additional here until that proves to not be sufficient.